LSPS5/README.md

```diff
@@ -0,0 +1,264 @@
+  * MUST treat a possible channel open, such as via a
+    reserved SCID for a JIT channel, as a channel open, for
+    purposes of remembering the `webhook` for a client.
+
+The `lsps5.setwebhook` call results in the empty object
+`{}` on success, and has the following errors defined:
+
+* `unsupported_version` (1) - The `version` specified by
+  the client is not among the `versions` that the LSP
+  supports.
+* `unsupported_protocol` (2) - The `webhook` is not an
+  HTTPS-protocol URI.
+
+### Webhook Call
+
+The LSP SHOULD make a `POST` request to the given webhook
+URI, if the client has registered one via `lsps5.setwebhook`
+*and* one of the following conditions is true:
+
+* The LSP learns of an incoming payment to the client
+  (whether an incoming HTLC, or other future mechanism
+  for incoming payments).
+* The LSP wants to take back some of its liquidity towards
+  the client.
+* The LSP wants to change feerates for any channels towards
+  the client, or feerates for any auto-management mechanism
+  the client has set up with the LSP.
+* The LSP wants to withdraw all services from the client
+  and close all channels with the client.
+
+The LSP makes a `POST` request whose content body has the
+MIME type `application/json`, containing a JSON object like
+the below:
+
+```JSON
+{
+  "version": 1,
+  "lsp": "023456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef012",
+  "details": {
+    "reason": 1,
+    "pushed_at": "2023-04-25T14:27:03.842Z",
+    "zbase": "blahblah"
+  }
+}
+```
+
+`version` is a required integral number, the version of the
+API that was specified in the `lsps5.setwebhook` call.
+
+`lsp` is a required node ID, the node ID of the LSP that is
+making this push request.
+
+`details` is a required object whose format is dependent on
+`version`.
+
+* For `version` 1, `details` has the fields:
+  * `reason` is a required integral number, which is one
+    of the reason codes defined below.
+  * `pushed_at` is a required datetime, which is the time
+    at the LSP when it generated this push request.
+  * `zbase` is a required signature from the LSP, signing
+    the message defined below.
+
+For `version` 1 `zbase` signatures, the signature signs
+the following JSON string:
+
+    "LSPS5:Version 1: Reason ${reason} at time ${pushed_at}."
+
+Where `${reason}` is the shortest decimal text form of the
+`reason` code, and `${pushed_at}` is the ISO8601 format
+`YYYY-mm-ddThh:mm:ss.uuuZ` of the `pushed_at` datetime.
+
+For example, for the above `version` 1 `details` object,
+the JSON string to be signed would be `"LSPS5:Version 1:
+Reason 1 at time 2023-04-25T14:27:03.842Z."`
+
+For `version` 1, the following `reason` codes are defined:
+
+* `1` - The client has a pending incoming payment.
+* `2` - The LSP wants to recover some of the liquidity from
+  the client.
+* `3` - The LSP wants to change feerates for channels or
+  other management actions.
+* `4` - The LSP wants to end services to the client.
```

Comment:
If we decide to use JSON-RPC 2.0 notifications, instead of a `reason` code we could just use `method` name.  E.g.

```JSON
{
  "jsonrpc": "2.0",
  "method": "lsps5.youhavenewpayments",
  "param": {
    "version": 1,
    "total_msat": "546000"
  }
}
```

Then put signature and time in HTTP header, as @SeverinAlexB proposed.

Reply:
Structure looks good. That's also a neat way to diversify method bodies if necessary.

I think I prefer the signature in the message itself. Just so it's easily portable to other protocols if necessary (like noise maybe?)

Reply:
If we go with signature-in-the-message we might want to do something like:

```JSON
{
  "jsonrpc": "2.0",
  "method": "youhavenewpayments",
  "params": {
    "signature": "blah",
    "time": "2023-05-01T18:48:03.581Z",
    "details": {
      "version": 1,
      "total_msat": "546000"
    }
  }
}
```

Then the signature signs `method`, `time`, and `details`.

Reply:
Quoting myself above:

> I don't see why we should support anything else than HTTP webhooks?

Reply:
Looks good. Either way, if you're going to sign json you'll need some strict rules on how that json should be formatted. No whitespaces, keys in alphabetical order, etc (maybe that was all).

Reply:
Not sure why you need any rules. You sign the raw http body. This can be whatever format.

Reply:
Indeed, if the signature is not inside the JSON itself, we could sign the raw HTTP body directly.

Reply:
Went with signature in the HTTP header.